### PR TITLE
algod importer: Auto catchpoint label lookup.

### PR DESCRIPTION
**Pull request by winder**
## Summary

Adds a new option to automatically lookup an appropriate catchpoint label during startup.

## Test Plan

New unit tests.

Manually run conduit with `-r 27000001` to see that the correct catchpoint label is automatically populated.

### CODE REVIEW

**Review by Eric-Warehime — APPROVED** [review]

**Review by tzaffi — COMMENTED** [review]
Still reading but it's a good juncture to submit some feedback.

File: conduit/plugins/importers/algod/README.md [hunk 1]
```diff
@@ -34,12 +34,16 @@ When using a follower node, ledger state delta objects are provided to the proce
...
     # Algod catchpoint catchup arguments
     catchup-config:
-        # The catchpoint to use when running fast catchup. Select an appropriate catchpoint for your deployment.
-        # They are published in the following locations:
-        # mainnet: https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/mainnet_catchpoints.txt
-        # betanet: https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/betanet_catchpoints.txt
-        # testnet: https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/testnet_catchpoints.txt
+        # Automatically download an appropriate catchpoint label. If false, you
+        # must specify a catchpoint to use fast catchup.
+        auto: false
```
[the remark this answers is not in the document]
Comment (winder) [reply conduit/plugins/importers/algod/README.md:39]: Good point. Which approach would you recommend? I updated the readme to clarify what was being done automatically. But maybe the setting name could also be clearer?

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md:39]: Thinking some more about it, another option is to use `manual` (instead of `auto`) and when this is omitted, it's zero-valued to `false` and the `catchpoint` field can be omitted as well. This assumes that users will mostly prefer non-manual catchup.

On a related front, I'm getting the below error when testing locally.  Should I have set the sync round _before_ starting conduit?

```
pipeline init error: Pipeline.Init(): could not initialize importer (algod): POST [/v2/catchup/26580000](https://file+.vscode-resource.vscode-cdn.net/v2/catchup/26580000)#VB7MWRDWQD3H4Z3A6R5XMIYXVMPSLQRVRVT7REW472NGRZNQGY7A received unexpected error: HTTP 400: {"message":"failed to start catchup : unable to start catchpoint catchup for '26580000#VB7MWRDWQD3H4Z3A6R5XMIYXVMPSLQRVRVT7REW472NGRZNQGY7A' - resulting round is beyond current sync round '1'"}
```

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md:39]: To answer your question, I'm leaning towards just modding the README. `auto` is a sensible name (or the opposite `manual` as well)

File: docs/tutorials/IndexerMigration.md [hunk 2]
```diff
@@ -255,11 +253,12 @@ metrics:
 log-level: "INFO"
 importer:
   name: "algod",
-  config: 
+  config:
     "netaddr": $ALGOD_ADDR,
     "token": $ALGOD_TOKEN,
     "mode": "follower",
     catchup-config:
+        auto: false
```
[the remark this answers is not in the document]
Comment (winder) [reply docs/tutorials/IndexerMigration.md:261]: I don't think so.

File: conduit/plugins/importers/algod/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md:39]: And to answer my own question, the following script seems to work. So yeah, you do need to call `sync` before starting conduit:

```jupyter
%%bash
export DATA_NODE=follower

just stop_and_nuke || true

just pub-create conduit mainnet

just start

just algod /v2/ledger/sync/26589217 POST
just algod /v2/ledger/sync

../cmd/conduit/conduit -d ./data --next-round-override 26589217
```

**Review by shiqizng — APPROVED** [review]

File: conduit/plugins/importers/algod/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (winder) [reply conduit/plugins/importers/algod/README.md:39]: When this is all finished, you shouldn't have to call sync manually. You might have an older version of this branch.

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md:39]: I'll try with the latest. Meanwhile, my local follower was stuck in the catchpoint phase and wasn't exporting any blocks. When I killed conduit, the follower immediately got back into a normal state.

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md:39]: Interesting. It might be an artifact of running conduit inside of Jupyter notebook. Now that I attempted to kill the process, a bunch of blocks are getting written. I'll run my next experiment in a vanilla terminal.

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md:39]: So I experimented a bit more as follows:
* pulled down 28487271 from https://github.com/algorand/go-algorand/pull/5349
* `make install`'ed it
* started a follower node using `goal` against `mainnet`
* tried _not_ setting the sync round AND also later tried setting the sync round to 10009217
* set the following in `conduit.yml`
  * retry-count: 1
  * retry-delay: "1s"
  * mode: "follower"
  * netaddr: FOLLOWER_LOCAL_URL
  * token: FOLLOWER_ALGOD_TOKEN
  * catchup-config:
    *  auto: true
    *  admin-token: FOLLOWER_ALGOD_ADMIN_TOKEN
* pulled down dcb0cde from this PR
* `make`d it
* `../cmd/conduit/conduit -d ./data --next-round-override 10009217`

## I. What happened when I didn't set the sync round before starting conduit?
Follower was taking a very long time to catchup. In fact, it didn't really catchup in time before conduit gave up and failed to load the block. 

Here is a typical algod status:
```
❯ just status
Last committed block: 225876
Time since last block: 0.0s
Sync Time: 952.5s
Last consensus protocol: https://github.com/algorandfoundation/specs/tree/5615adc36bad610c7f165fa2967f4ecfa75125f0
Next consensus protocol: https://github.com/algorandfoundation/specs/tree/5615adc36bad610c7f165fa2967f4ecfa75125f0
Round for next consensus protocol: 225877
Next consensus protocol supported: true
Last Catchpoint: 
Genesis ID: mainnet-v1.0
Genesis hash: wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=
```

And here are the last logs of conduit:
```
{"__type":"processor","_name":"algod","level":"error","msg":"error getting block for round 10009217 (attempt 4)","time":"2023-05-03T22:01:44-05:00"}
{"__type":"processor","_name":"algod","level":"error","msg":"failed to get block for round 10009217 after 5 attempts, check node configuration: HTTP 404: {\"message\":\"failed to retrieve information from the ledger\"}\n","time":"2023-05-03T22:01:44-05:00"}
{"__type":"Conduit","_name":"main","level":"error","msg":"failed to get block for round 10009217 after 5 attempts, check node configuration: HTTP 404: {\"message\":\"failed to retrieve information from the ledger\"}\n","time":"2023-05-03T22:01:44-05:00"}
{"__type":"Conduit","_name":"main","level":"error","msg":"Pipeline has exceeded maximum retry count (1) - stopping...","time":"2023-05-03T22:01:44-05:00"}
{"__type":"processor","_name":"file_writer","level":"info","msg":"latest round on file: 10009217","time":"2023-05-03T22:01:44-05:00"}

Exiting with error:
failed to get block for round 10009217 after 5 attempts, check node configuration: HTTP 404: {"message":"failed to retrieve information from the ledger"}
```

## II. What happened when I didn't set the sync round to 10009217 before starting conduit?
Basically the same as not setting the sync round.
About a minute in conduit announces that it finished the catchup, while the follower is really only on round 20,000. After about 6 minutes conduit gives up attempting to retrieve the round.

[the remark this answers is not in the document]
Comment (winder) [reply conduit/plugins/importers/algod/README.md]: Sorry for the confusion. I've figured out the problem. You need the changes in [go-algorand#5349](https://github.com/algorand/go-algorand/pull/5349). Without them you run into this error: https://github.com/algorand/go-algorand/pull/5349/files#r1185050820

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md]: Hmmm... I did pull down and make the latest from 5349. When I run:
```
❯ goal --version
12886130966
3.16.180502.dev [HEAD] (commit #28487271+)
go-algorand is licensed with AGPLv3.0
source code available at https://github.com/algorand/go-algorand
```
and git log gives:
```
commit 284872714e46ae7504e02a7288c9837a000942a3 (HEAD, will/will/follower-catchup-pause)
Author: Will Winder <wwinder.unh@gmail.com>
Date:   Wed May 3 07:19:57 2023 -0400

    Remove old test.
```
So are the errors I'm seeing necessarily from your link?

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md]: One possible issue is that I'm merging into another branch. I'll try a clean version of the commit.

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md]: Ran again with a clean version of this PR. Taking some notes:

Command:
```sh
❯ conduit -d ./data2 -r 10009217
... after around 30 seconds ...
{"__type":"Conduit","_name":"main","level":"info","msg":"Pipeline round: 10009217","time":"2023-05-04T10:25:38-05:00"}
... another minute or so ...
{"__type":"processor","_name":"algod","level":"error","msg":"error getting block for round 10009217 (attempt 0)","time":"2023-05-04T10:26:38-05:00"}
{"__type":"processor","_name":"algod","level":"error","msg":"error getting block for round 10009217 (attempt 1)","time":"2023-05-04T10:27:38-05:00"}
```

at the same time getting algod's status (30s and around 3 minutes in)
```
❯ just status
Last committed block: 6483
Time since last block: 0.0s
Sync Time: 33.6s
Last consensus protocol: https://github.com/algorandfoundation/specs/tree/5615adc36bad610c7f165fa2967f4ecfa75125f0
Next consensus protocol: https://github.com/algorandfoundation/specs/tree/5615adc36bad610c7f165fa2967f4ecfa75125f0
Round for next consensus protocol: 6484
Next consensus protocol supported: true
Last Catchpoint: 
Genesis ID: mainnet-v1.0
Genesis hash: wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=
RUNNING
❯ just status
Last committed block: 39148
Time since last block: 0.0s
Sync Time: 225.6s
Last consensus protocol: https://github.com/algorandfoundation/specs/tree/5615adc36bad610c7f165fa2967f4ecfa75125f0
Next consensus protocol: https://github.com/algorandfoundation/specs/tree/5615adc36bad610c7f165fa2967f4ecfa75125f0
Round for next consensus protocol: 39149
Next consensus protocol supported: true
Last Catchpoint: 
Genesis ID: mainnet-v1.0
Genesis hash: wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=
RUNNING
```

[the remark this answers is not in the document]
Comment (tzaffi) [reply conduit/plugins/importers/algod/README.md]: Ended up failing the same way as before.
